### PR TITLE
Add retrieval function for BSRN data

### PR DESCRIPTION
**Pull request by AdamRJensen**
- [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Add a retrieval function for BSRN data `pvlib.iotools.get_bsrn`. Data is obtained from the BSRN's FTP server.

### CODE REVIEW

**Comment by AdamRJensen** [comment]
@kanderso-nrel This is my initial idea for the `get_bsrn` function. 

The `read_bsrn` function needs to be updated to take a file buffer or a `parse_bsrn` function needs to be created.

**Comment by wholmgren** [comment]
@AdamRJensen our [solarforecastarbiter-core adaptation](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/solarforecastarbiter/io/fetch/bsrn.py) of the BSRN code that you contributed to pvlib includes a change to make a `parse_bsrn` function that only takes a file buffer. Maybe that would be helpful to copy back over here.

**Comment by AdamRJensen** [comment]
@kanderso-nrel Any idea of why the Linux versions fail, but the others pass?

The log states that the test `test_get_bsrn_no_files` is failing on Linux due to "Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) was emitted." 

My knowledge of Linux is too limited to understand why it would fail on Linux and not the others.. any ideas?

**Comment by kanderso-nrel** [comment]
> @kanderso-nrel Any idea of why the Linux versions fail, but the others pass?

I don't think it's actually a linux-specific issue, it's just that only the `conda_linux` versions are configured to run the `remote_data` tests.  I can recreate the CI failure on a local ubuntu VM.  The problem is that it actually does find a file for the bogus date -- I get back data for the month of `2000-06`.  I guess the issue is that files are named with the format `f"{station}%m%y.dat.gz"`, so year 1800 gets formatted as year 00 and interpreted as year 2000.  Changing the `start` and `end` from 1800 to 1899 gets the test to pass for me.

**Comment by AdamRJensen** [comment]
@cwhanse Any opinion on adding **"gri" (ground-reflected irradiance)** to the pvlib list of variables? Better alternatives for the naming of *upward short-wave reflected irradiance*?

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:156]: When passing `start=end=2019-01-31`, data for both 2019-01 and 2019-02 are returned.  It's a little clunky, but what do you think about `pd.date_range(start.replace(day=1), end + pd.tseries.offsets.MonthEnd(n=0), freq='1M')`?

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:186]: There's a lot of shuffling data to/from buffers here (note that `gzip.decompress` wraps the input in a buffer as well).  I wonder if it might be cleaner to use `gzip.GzipFile` directly, maybe with an `io.TextIOWrapper`?  Maybe not, and probably this is a small fraction of total runtime anyway, so maybe not worth optimizing.

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:256]: More for my own education: what does the second `strip()` accomplish here?

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:262]: The BSRN's lat/lon convention isn't consistent with pvlib's.  Should we convert it?  As always I'm torn between making things convenient for pvlib users vs faithfully reporting whatever the provider gives us.

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:281]: What's the reason for `seek`ing back to the beginning here?  I think the data records will always be after `LR0004`, so is there any problem with just mapping the sections that appear after `LR0004`?

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:287]:
```suggestion
                lr_nrows[lr] = num - lr_startrow[lr]
```
maybe clearer?  might need a -1 or something too

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:220]: Maybe I'm doing it wrong but I don't think `str` is actually allowed here (or the other functions)

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:295]:
```suggestion
                             "['0100', '0300', '0500'].")
```

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:384]:
```suggestion
        returned if the specified logical records were not found.

```

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:408]:
```suggestion
    lwu†                     float   Mean long-wave upwelling irradiance [W/m^2]
```

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [inline pvlib/iotools/bsrn.py:426]:
```suggestion
    According to [2]_ "All time labels in the station-to-archive files denote
```

File: pvlib/tests/iotools/test_bsrn.py
Comment (kanderso-nrel) [inline pvlib/tests/iotools/test_bsrn.py:17]:
```suggestion
    Users should obtain their own credentials as described in the `read_bsrn`
```

File: pvlib/tests/iotools/test_bsrn.py
Comment (kanderso-nrel) [inline pvlib/tests/iotools/test_bsrn.py:64]: seems like this comment doesn't match the actual test

**Review by kanderso-nrel — COMMENTED** [review]
This looks great @AdamRJensen!  Comments below, mostly very nitpicky...

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:256]: Well, the second strip() is redundant. But strip() removes starting and trailing whitespaces, as well as "\r\n" and "\n"

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:281]: If the seek is remove here, then we need to manually add the line number offset (where the metadata section ends) to `enumerate` using *start* as otherwise, the line numbering will be incorrect in the dictionary.

As the metadata section and the earlier sections have a variable number of lines this will probably be rather complex to keep track of.

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:156]: what about just:
`pd.date_range(start.replace(day=1), end.replace(day=1) + pd.DateOffset(months=1), freq='1M')`

I believe you might not even need the `replace()` on start.

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:63]: I actually have it right now so it can either take a list or string. What do you think of that?
`    logical_records: str or list, default: ['0100']`

Therefore, the default could just be the string version: '0100'.

Otherwise, should I also change the input type in the doc string to tuple?

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:220]: In the parse function I have this line, which allows it to be string:
`    for lr in list(logical_records):`

This does however also have to be added to the empty dataframe function.

First though, is this a desireable functionality or should I remove this?

Comment (kanderso-nrel) [reply pvlib/iotools/bsrn.py:220]: Note that `list('0100') == ['0', '1', '0', '0']`.  Does it work for you if you pass `logical_records='0100'`?  I get `ValueError: Logical record 0 not in ['0100', '0300', '0500'].`  I'd say just require it to be a list (or tuple etc).

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:426]: Sorry for all the small typos...

File: pvlib/iotools/bsrn.py
Comment (kanderso-nrel) [reply pvlib/iotools/bsrn.py:156]: Works for me.  I wasn't sure about the `replace()` on start, I included it to be safe but I think you're right that it's unnecessary.

File: pvlib/tests/iotools/test_bsrn.py
Comment (AdamRJensen) [reply pvlib/tests/iotools/test_bsrn.py:64]: What makes you think that?

In line 67 of the same test, the index is checked if it is empty:
`    assert_index_equal(pd.DataFrame().index, data.index)`

Perhaps that I check for certain columns to be present mislead you. That is only to check that the empty dataframe has the correct columns.

Comment (kanderso-nrel) [reply pvlib/tests/iotools/test_bsrn.py:64]: > What makes you think that?

Careless reading.  I retract my objection :)

`assert data.empty` might be a little easier to parse than `assert_index_equal(pd.DataFrame().index, data.index)`

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:220]: You are indeed correct! List or tuples it is :)

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:287]: Nice tough. Certainly makes it much clearer. The -1 is needed fyi.

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:262]: There certainly are pros and cons. Either way it would be nice if it was specified, perhaps one way of doing this is simply to have two entries for latitude and longitude.

E.g., `metadata['latitude']` would be the pvlib convention and `metadata['latitude_bsrn_convention']` would be the BSRN method.

File: pvlib/tests/iotools/test_bsrn.py
Comment (AdamRJensen) [reply pvlib/tests/iotools/test_bsrn.py:64]: Hahah. I have changed it to `data.empty`, which probably would have avoided the confusion.

File: pvlib/iotools/bsrn.py
Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:186]: I took a lot of fiddling around to find this solution, as I was stumbling into errors for all the other solutions I investigated.

When I run `gzip.GzipFile(bio)`, I get the following error:
`TypeError: expected str, bytes or os.PathLike object, not _io.BytesIO`

Comment (kanderso-nrel) [reply pvlib/iotools/bsrn.py:186]: Right, it would use the `fileobj` parameter of `GzipFile`, I think this might work:

```python
bio.seek(0)
gzip_file = io.TextIOWrapper(gzip.GzipFile(fileobj=bio), encoding='utf-8')  # or latin1?
dfi, metadata = parse_bsrn(gzip_file, logical_records)
```

Again, not sure this is worth worrying about though.

Comment (AdamRJensen) [reply pvlib/iotools/bsrn.py:186]: My solution takes 83 ms, whereas yours take 20.6 µs. So certainly a bit of a performance gain :)

I'll change it to your solution

**Comment by AdamRJensen** [comment]
Thanks for the review @kanderso-nrel! Do we need a second review on this?

**Comment by kanderso-nrel** [comment]
> Do we need a second review on this?

My answer to that question is always yes :) https://en.wikipedia.org/wiki/Linus%27s_law

**Review by kanderso-nrel — APPROVED** [review]
LGTM, @pvlib/pvlib-maintainer anyone else want to review this?

**Comment by cwhanse** [comment]
> LGTM, @pvlib/pvlib-maintainer anyone else want to review this?

I don't see how I will make time for it. I vote merge and improve later, if needed.

**Review by wholmgren — APPROVED** [review]
Looks good to me. I confirmed the parse function works with data I obtained from the NASA Langley site using requests.

The lines that are not covered are either expected due to the test configuration or are ok to skip for iotools.

Thanks @AdamRJensen! I expect this will be useful to many people.

**Pull request merged**
